### ISSUE TITLE
`SharedTokenCacheCredential()` fails with unexpected error causing `DefaultAzureCredential()` to fail

### ISSUE THREAD

**Issue by natbprice** [Azure/azure-sdk-for-python#24032]
- **Package Name**: azure-identity
- **Package Version**: 1.9.0
- **Operating System**: Windows 10
- **Python Version**: 3.6.13

**Describe the bug**
`SharedTokenCacheCredential()` fails with unexpected error causing `DefaultAzureCredential()` to not proceed to next authentication method (e.g., `VisualStudioCodeCredential()` or `AzureCliCredential()`)

**To Reproduce**
1. Create an invalid token in the local cache shared between Microsoft applications (change user password?)
2. Pass `DefaultAzureCredential()` to supported function
3. Observe that authentication chain fails at `SharedTokenCacheCredential()` without continuing to next authentication method

```
INFO:azure.identity.aio._credentials.environment:No environment configuration found.
INFO:azure.identity.aio._credentials.managed_identity:ManagedIdentityCredential will use IMDS
WARNING:azure.identity.aio._internal.decorators:EnvironmentCredential.get_token failed: EnvironmentCredential authentication unavailable. Environment variables are not fully configured.
INFO:azure.identity.aio._credentials.chained:DefaultAzureCredential - EnvironmentCredential is unavailable
INFO:azure.core.pipeline.policies.http_logging_policy:Request URL: 'http://169.254.169.254/metadata/identity/oauth2/token'
INFO:azure.core.pipeline.policies.http_logging_policy:Request method: 'GET'
INFO:azure.core.pipeline.policies.http_logging_policy:Request headers:
INFO:azure.core.pipeline.policies.http_logging_policy:    'Metadata': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'User-Agent': 'azsdk-python-identity/1.5.0 Python/3.9.1 (Windows-10-10.0.19041-SP0)'
INFO:azure.core.pipeline.policies.http_logging_policy:No body was attached to the request
INFO:azure.core.pipeline.policies.http_logging_policy:Response status: 400
INFO:azure.core.pipeline.policies.http_logging_policy:Response headers:
INFO:azure.core.pipeline.policies.http_logging_policy:    'Cache-Control': 'no-store, no-cache'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Pragma': 'no-cache'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Content-Type': 'application/json; charset=utf-8'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Expires': '-1'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Strict-Transport-Security': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'X-Content-Type-Options': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'P3P': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'x-ms-request-id': '67f3f25b-98bf-4945-ae6b-d4e300f94400'
INFO:azure.core.pipeline.policies.http_logging_policy:    'x-ms-ests-server': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'X-XSS-Protection': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Set-Cookie': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Set-Cookie': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Set-Cookie': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Date': 'Fri, 15 Apr 2022 17:56:19 GMT'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Content-Length': '737'
WARNING:azure.identity.aio._internal.decorators:SharedTokenCacheCredential.get_token failed: Azure Active Directory error '(invalid_grant) AADSTS50173: The provided grant has expired due to it being revoked, a fresh auth token is needed. The user might have changed or reset their password. The grant was issued on '2022-02-15T15:21:36.6536759Z' and the TokensValidFrom date (before which tokens are not valid) for this user is '2022-04-03T18:41:59.0000000Z'.
Trace ID: 67f3f25b-98bf-4945-ae6b-d4e300f94400
Correlation ID: 4a7daacb-b66d-4b35-be80-36414066a315
Timestamp: 2022-04-15 17:56:20Z'
WARNING:azure.identity.aio._credentials.chained:DefaultAzureCredential.get_token failed: SharedTokenCacheCredential raised unexpected error "Azure Active Directory error '(invalid_grant) AADSTS50173: The provided grant has expired due to it being revoked, a fresh auth token is needed. The user might have changed or reset their password. The grant was issued on '2022-02-15T15:21:36.6536759Z' and the TokensValidFrom date (before which tokens are not valid) for this user is '2022-04-03T18:41:59.0000000Z'.
Trace ID: 67f3f25b-98bf-4945-ae6b-d4e300f94400
Correlation ID: 4a7daacb-b66d-4b35-be80-36414066a315
Timestamp: 2022-04-15 17:56:20Z'"
INFO:azure.identity.aio._credentials.managed_identity:No response from the IMDS endpoint.
WARNING:azure.identity.aio._internal.decorators:ManagedIdentityCredential.get_token failed: ManagedIdentityCredential authentication unavailable, no managed identity endpoint found.
INFO:azure.identity.aio._credentials.chained:DefaultAzureCredential - ManagedIdentityCredential is unavailable
INFO:azure.core.pipeline.policies.http_logging_policy:Request URL: 'https://login.microsoftonline.com/organizations/oauth2/v2.0/token'
INFO:azure.core.pipeline.policies.http_logging_policy:Request method: 'POST'
INFO:azure.core.pipeline.policies.http_logging_policy:Request headers:
INFO:azure.core.pipeline.policies.http_logging_policy:    'Content-Type': 'application/x-www-form-urlencoded'
INFO:azure.core.pipeline.policies.http_logging_policy:    'User-Agent': 'azsdk-python-identity/1.5.0 Python/3.9.1 (Windows-10-10.0.19041-SP0)'
INFO:azure.core.pipeline.policies.http_logging_policy:A body is sent with the request
INFO:azure.core.pipeline.policies.http_logging_policy:Response status: 400
INFO:azure.core.pipeline.policies.http_logging_policy:Response headers:
INFO:azure.core.pipeline.policies.http_logging_policy:    'Cache-Control': 'no-store, no-cache'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Pragma': 'no-cache'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Content-Type': 'application/json; charset=utf-8'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Expires': '-1'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Strict-Transport-Security': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'X-Content-Type-Options': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'P3P': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'x-ms-request-id': '55c2175c-28d5-48d0-b050-e3c2216f6500'
INFO:azure.core.pipeline.policies.http_logging_policy:    'x-ms-ests-server': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'X-XSS-Protection': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Set-Cookie': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Set-Cookie': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Set-Cookie': 'REDACTED'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Date': 'Fri, 15 Apr 2022 17:56:30 GMT'
INFO:azure.core.pipeline.policies.http_logging_policy:    'Content-Length': '737'
WARNING:azure.identity.aio._internal.decorators:SharedTokenCacheCredential.get_token failed: Azure Active Directory error '(invalid_grant) AADSTS50173: The provided grant has expired due to it being revoked, a fresh auth token is needed. The user might have changed or reset their password. The grant was issued on '2022-02-15T15:21:36.6536759Z' and the TokensValidFrom date (before which tokens are not valid) for this user is '2022-04-03T18:41:59.0000000Z'.
Trace ID: 55c2175c-28d5-48d0-b050-e3c2216f6500
Correlation ID: f1dc4b03-e95a-4747-bdab-08df9d2dc7fa
Timestamp: 2022-04-15 17:56:31Z'
WARNING:azure.identity.aio._credentials.chained:DefaultAzureCredential.get_token failed: SharedTokenCacheCredential raised unexpected error "Azure Active Directory error '(invalid_grant) AADSTS50173: The provided grant has expired due to it being revoked, a fresh auth token is needed. The user might have changed or reset their password. The grant was issued on '2022-02-15T15:21:36.6536759Z' and the TokensValidFrom date (before which tokens are not valid) for this user is '2022-04-03T18:41:59.0000000Z'.
Trace ID: 55c2175c-28d5-48d0-b050-e3c2216f6500
Correlation ID: f1dc4b03-e95a-4747-bdab-08df9d2dc7fa
Timestamp: 2022-04-15 17:56:31Z'"
 Error in py_call_impl(callable, dots$args, dots$keywords) : 
  AttributeError: 'NoneType' object has no attribute 'headers'
```

**Expected behavior**
- If `SharedTokenCredential()` is last method in chain, a descriptive error message is returned (not `AttributeError: 'NoneType' object has no attribute 'headers'`)
- If `SharedTokenCredential()` is not the last method in chain, `DefaultAzureCredential()` and `ChainedTokenCredential()` proceed to next method
- Warning message includes information about how to clear cache and/or refresh token

**Additional context**
Related issue #21416

**Comment by mccoyp** [comment]
Hi again @natbprice, thank you for opening this issue, tagging the related issue, and providing logs! I'll tag someone who should be able to help and we'll get back to you as soon as possible.

**Comment by mccoyp** [comment]
@natbprice the error you're getting here seems to be caused by the cached token being invalid, due to having permissions revoked, per the error message. This is slightly different than the issue described in https://github.com/Azure/azure-sdk-for-python/issues/21718 (where a refresh token had expired), but a workaround should be the same in your case: excluding the `SharedTokenCacheCredential` from the `DefaultAzureCredential` chain.

You can pass the `exclude_shared_token_cache_credential` option to the credential constructor to skip this credential (`DefaultAzureCredential(exclude_shared_token_cache_credential=True)`). Does this work to resolve the immediate blocker?

**Comment by natbprice** [comment]
Thanks @mccoyp for responding so quickly! I see now that #21718 is very similar to the issue I reported here. It seems the solution is to simply exclude `SharedTokenCacheCredential()`. Also, that issue discusses the logic behind why `DefaultAzureCredential()` should abort the search for other credentials when cache credentials do not work (https://github.com/Azure/azure-sdk-for-python/issues/21718#issuecomment-974225195). 

Therefore, maybe this issue as it relates to `azure-identity` has already been addressed and we can close this ticket.

It would be nice if the error message was more informative, but I think you were suggesting that would be addressed separately in #21416 as it relates to `azure-data-tables`.

**Comment by mccoyp** [comment]
Thank you @natbprice for bearing with the multiple tickets and responding quickly as well!

Yes, the short term solution to unblock the `DefaultAzureCredential` is to skip credentials that have a stale cached token (as #21718 shows, `VisualStudioCodeCredential` can also be affected by this). `azure-identity` could probably make the error message in this scenario more descriptive and point to this solution; #21788 tracks this effort, and I'll look into describing this in the [`azure-identity` troubleshooting guide](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/identity/azure-identity/TROUBLESHOOTING.md).

The high-level error that you're seeing from `azure-data-tables` does obscure the underlying auth issue, and you're correct that #21416 will track fixing this. Thank you for the feedback and for reviving that thread!

It sounds like this ticket can be marked as resolved, but feel please feel free to update this if that's not the case.
